[DPE-5409] docs: update mTLS docs for Kafka 3 (#350)
canonical/kafka-operator: docs/how-to/create-mtls-client-credentials.md

(how-to-create-mtls-client-credentials)=
# Use mTLS for clients

Requirements:

- Charmed Apache Kafka cluster up and running
- [Encryption enabled](how-to-enable-encryption)
- [{spellexception}`Java Runtime Environment (JRE)`](https://ubuntu.com/tutorials/install-jre#1-overview) installed
- [`charmed-kafka` snap](https://snapcraft.io/charmed-kafka) installed
- [jq](https://snapcraft.io/jq) installed

This guide includes step-by-step instructions on how to create mTLS credentials for a client application to be able to connect to a Charmed Apache Kafka cluster.

## Create mTLS client credentials

> Each Apache Kafka mTLS client needs its own TLS certificate, which should be trusted by the charmed Apache Kafka application. In a typical production environment, certificates are issued either by the organization's PKI infrastructure, or trusted Certificate Authorities (CAs). For the sake of this guide, we will generate self-signed certificates.

Reviewer: chore: Please make sure the spellcheck test is green. You can run it locally with `make spelling` from the `docs` folder.

Author: Done, I love these new docs checks :)

  CSqGSIb3DQEBCwUAA4IBAQBSe2nUHoLA5Snn7R+r/Jp+agBjFAHT0LslULG0z7/s
  GCo2/W84q2KOlDP1kUJ0C6JBeS1BZTd/ZzAuHBmCWkwOOmQnHCvzT3vxdLuKOab1
  tdGZg0nvxiU1FDSkTchMccmeUMRk3aKzrYUMNg2PLxl2u0GdJhmYjhtARUte3nzo
  ufPtyMx80lOEK02O8vzgwYidVr7xplbg8SdLKbGwMLH03Wv3w7ew9kN743HbT8AM
  nx4xSLCybz3upJpRMXXkIenf7Rr7eDp3s4deAbGcvCo6B/XDeBOkSv8Sl7CsSejm
  05qe06cC/6/K45CzOrWRwr4q5m6ENK/UT5fuOLFIoVJS
  -----END CERTIFICATE-----
```

Copy the certificate content into a file named `server.pem` and save it. You can also do that using a single command:

```bash
juju run self-signed-certificates/0 get-ca-certificate --format json | jq -r '."self-signed-certificates/0"."results"."ca-certificate"' > server.pem
```

## Create mutual trust relation: client -> server

Depending on the type of the client application, there might be different ways to trust. In this guide, we are using the console-based apps shipped with the `charmed-kafka` snap, which depend on java keystore/trustsores. Follow the steps below to create necessary Java keystore and truststore artifacts for the client application:

### Create client's keystore

Create the client's keystore using the following commands:

```bash

> # Java keystore and truststore passwords

Reviewer: Suggestion: Since we are editing this document, can we please make sure not to use comments in code blocks unless it's absolutely necessary? I feel like all comments in code blocks from this page can be placed before each respectable code block as instructions (potentially splitting a code block into multiple smaller ones). That is especially relevant for `bash`/`shell` type code blocks.

keytool -list -keystore client.truststore.jks -storepass $KAFKA_CLIENT_TRUSTSTORE_PASSWORD -v | grep until
```

## Define the client's credentials on the Apache Kafka cluster

Since you are using TLS certificates for authentication, you need to provide a way to map the client's certificate to usernames defined on the Apache Kafka cluster.

In charmed Apache Kafka, this could be done using the `ssl_principal_mapping_rules` config option, which defines how the certificate's common name is translated into a username, using a handy regex syntax (refer to [Apache Kafka's official documentation](https://kafka.apache.org/documentation/#security_authz_ssl) for more details on the syntax):

```bash
# Map the CN on the cert to be considered the principal (username)
juju config kafka ssl_principal_mapping_rules='RULE:^.*[Cc][Nn]=([a-zA-Z0-9\.-]*).*$/$1/L,DEFAULT'
```

This command will trigger a rolling restart of the charmed Apache Kafka application. Once the application settles to `active|idle` status, you can proceed to the next step.

## Add authorisation rules via ACLs for the client

> Grant read and write privileges to the mTLS client user over _group_, _topic_ and _transaction_ resources:

Reviewer: ```suggestion
Grant read and write privileges to the mTLS client user over` _group_`, `_topic_`, and `_transaction_` resources:
```

> ```bash

Reviewer: Suggestion: This is a huge code block with many instructions hidden as comments. We should avoid doing that, unless it's a file (like a python script). Can we please split it into multiple smaller code blocks with comments serving as text in-between them?

Author: Broke it to smaller parts here: https://github.com/canonical/kafka-operator/pull/350/commits/ced2e095c13f8c9605896a6c3bcadb316a536e18

(how-to-create-mtls-client-credentials)=
# Use mTLS for clients

Requirements:

- Charmed Apache Kafka cluster up and running
- [Encryption enabled](how-to-enable-encryption)
- [{spellexception}`Java Runtime Environment (JRE)`](https://ubuntu.com/tutorials/install-jre#1-overview) installed
- [`charmed-kafka` snap](https://snapcraft.io/charmed-kafka) installed
- [jq](https://snapcraft.io/jq) installed

This guide includes step-by-step instructions on how to create mTLS credentials for a client application to be able to connect to a Charmed Apache Kafka cluster.

## Create mTLS client credentials

> Each Apache Kafka mTLS client needs its own TLS certificate, which should be trusted by the charmed Apache Kafka application. In a typical production environment, certificates are issued either by the organisation's PKI infrastructure, or trusted Certificate Authorities (CAs). For the sake of this guide, we will generate self-signed certificates.

Reviewer 2: @izmalk - When are we permitted to be conversational?

To generate a self-signed certificate in one prompt, use the following command:

```bash
openssl req -new -newkey rsa:4096 -days 365 -nodes -x509 -keyout client.key -out client.pem -subj "/C=US/ST=Denial/L=Springfield/O=Dis/CN=TestClient"
```

## Create mutual trust relation: server -> client

In order for the mTLS client to be able to communicate with the server (broker), the client should trust the broker's identity, and the broker should trust the client's identity. First, create the trust relation between the broker and the client.

> To trusts client certificates, the Apache Kafka charm's `trusted-certifcate` relation interface could be used. In order to integrate with that, you can deploy the `tls-certificates-operator` application and configure it to use the generated client certificate.

Reviewer 2: **todo:**
```suggestion
To trust client certificates, the `trusted-certifcate` relation interface is be used. Deploy the `tls-certificates-operator` application and configure it to use the generated client certificate:
```

> To trusts client certificates, the Apache Kafka charm's `trusted-certifcate` relation interface could be used. In order to integrate with that, you can deploy the `tls-certificates-operator` application and configure it to use the generated client certificate.

Reviewer: I like it, but we need to fix `is be used`. I think it's either `is to be used` or `is used`.

```suggestion
To trust client certificates, the `trusted-certifcate` relation interface is to be used. Deploy the `tls-certificates-operator` application and configure it to use the generated client certificate:
```

```

### Create client's truststore

Trust the broker's CA certificate by importing it into a Java truststore using the following command:

```bash
keytool -keystore client.truststore.jks -storepass $KAFKA_CLIENT_TRUSTSTORE_PASSWORD -noprompt \
  -importcert -alias server -file server.pem
```

### Check certificates validity

You can list the certificates loaded into client's keystore and truststore using the following commands:

```bash
echo "Client certs in Keystore:"
keytool -list -keystore client.keystore.p12 -storepass $KAFKA_CLIENT_KEYSTORE_PASSWORD -rfc | grep "Alias name"
keytool -list -keystore client.keystore.p12 -storepass $KAFKA_CLIENT_KEYSTORE_PASSWORD -v | grep until

echo "Server certs in Truststore:"
keytool -list -keystore client.truststore.jks -storepass $KAFKA_CLIENT_TRUSTSTORE_PASSWORD -rfc | grep "Alias name"
keytool -list -keystore client.truststore.jks -storepass $KAFKA_CLIENT_TRUSTSTORE_PASSWORD -v | grep until
```

## Define the client's credentials on the Apache Kafka cluster

Since you are using TLS certificates for authentication, you need to provide a way to map the client's certificate to usernames defined on the Apache Kafka cluster.

> In charmed Apache Kafka, this could be done using the `ssl_principal_mapping_rules` configuration option, which defines how the certificate's common name is translated into a username, using a handy regex syntax (refer to [Apache Kafka's official documentation](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=89071740) for more details on the syntax):

Reviewer: ```suggestion
In Charmed Apache Kafka, this is done using the `ssl_principal_mapping_rules` configuration option, which defines how the certificate's common name is translated into a username, using a regex (see [Apache Kafka's official documentation](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=89071740) for more details on the syntax):
```

```bash
juju config kafka ssl_principal_mapping_rules='RULE:^.*[Cc][Nn]=([a-zA-Z0-9\.-]*).*$/$1/L,DEFAULT'
```

This command will trigger a rolling restart of the charmed Apache Kafka application. Once the application settles to `active|idle` status, you can proceed to the next step.

## Add authorisation rules via ACLs for the client

> In order to add authorisation rules for the mTLS client, first save the broker's connection information and configuration path into some environment variables:

Reviewer: ```suggestion
To add authorisation rules for the mTLS client, first save the broker's connection information and configuration path into some environment variables:
```

```bash
BROKER_IP=$(juju show-unit kafka/0 --format json | jq -r '."kafka/0"."public-address"')
KAFKA_SERVERS_SASL="$BROKER_IP:19093"
KAFKA_SERVERS_MTLS="$BROKER_IP:9094"
SNAP_KAFKA_PATH=/var/snap/charmed-kafka/current/etc/kafka
```

Next, create the `KAFKA_CLIENT_MTLS_CN` environment variable holding client's certificate common name, this should be all lower-case because of the L suffix in the `ssl_principal_mapping_rules` configured before:

```bash
KAFKA_CLIENT_MTLS_CN=testclient

> ```

Reviewer 2: **todo:** Let's add a single line here saying something like '`testclient` is what the actual kafka username will be'. Just making it a little more explicit why we're doing the next step of granting access to `User:testclient`

> Finally, grant read and write privileges to the mTLS client user over `_group_`, `_topic_` and `_transaction_` resources using the following commands:

Reviewer: ```suggestion
Finally, grant read and write privileges to the mTLS client user over `_group_`, `_topic_` and `_transaction_` resources:
```

> Finally, grant read and write privileges to the mTLS client user over `_group_`, `_topic_` and `_transaction_` resources using the following commands:

Reviewer 2: **todo:**
```suggestion
Finally, grant read and write privileges to the mTLS client user over `--group`, `--topic` and `--transactional-id` resources using the following commands:
```

Author: Changed here: https://github.com/canonical/kafka-operator/pull/350/commits/cdb78279a3a5ff5452338e23e38f3c1c121b929f

## Test access

> To test the client's access, first create a file called `client-mtls.properties` with the following configuration:

Reviewer: Suggestion: Given the `cat` command, it seems like the following is the command to create the file, rather than the contents of the file. Correct me if I'm wrong.

```suggestion
To test the client's access, first create a file called `client-mtls.properties`:
```

```bash
cat <<EOF > client-mtls.properties
security.protocol=SSL
bootstrap.servers=$KAFKA_SERVERS_MTLS
ssl.truststore.location=$SNAP_KAFKA_PATH/client.truststore.jks
ssl.truststore.password=$KAFKA_CLIENT_TRUSTSTORE_PASSWORD
ssl.truststore.type=JKS
ssl.keystore.location=$SNAP_KAFKA_PATH/client.keystore.p12
ssl.keystore.password=$KAFKA_CLIENT_KEYSTORE_PASSWORD
ssl.keystore.type=PKCS12
ssl.client.auth=required
EOF
```

Next, copy the files to a path readable by the `charmed-kafka` snap commands:

```bash
sudo cp client.truststore.jks $SNAP_KAFKA_PATH/
sudo cp client.keystore.p12 $SNAP_KAFKA_PATH/
sudo cp client-mtls.properties $SNAP_KAFKA_PATH/
```

> Apply file permissions to the copied files so that they are readable by the snap:

Reviewer: Question: IIRC, the `chown` command is about ownership, rather than permissions. Can we make the description more accurate here?

Author: Thanks you're right. Changed it here: https://github.com/canonical/kafka-operator/pull/350/commits/cdb78279a3a5ff5452338e23e38f3c1c121b929f
